docs/community/tutorials/bert-qa-benchmarking/index-bert-qa-benchmarking-with-redisai-and-redisgears.mdx

```diff
@@ -0,0 +1,406 @@
+
+![featured](featured.jpg)
+
+# BERT Question Answering pipeline and API
+
+In the BERT QA pipeline (or in any other modern NLP inference task), there are two steps:
+
+1. Tokenize text - turn text into numbers
+2. Run the inference - large matrix multiplication
+   With Redis, we have the opportunity to pre-compute everything and store it in memory, but how do we do it?
+   Unlike with summarization ML learning task, the question is not known in advance, so can't pre-compute all possible answers. But we can pre-tokenize all potential answers - all paragraphs in the dataset using RedisGears:
+
+```python
+def parse_sentence(record):
+    import redisAI
+    import numpy as np
+    global tokenizer
+    if not tokenizer:
+        tokenizer=loadTokeniser()
+    hash_tag="{%s}" % hashtag()
+
+    for idx, value in sorted(record['value'].items(), key=lambda item: int(item[0])):
+        tokens = tokenizer.encode(value, add_special_tokens=False, max_length=511, truncation=True, return_tensors="np")
+        tokens = np.append(tokens,tokenizer.sep_token_id).astype(np.int64)
+        tensor=redisAI.createTensorFromBlob('INT64', tokens.shape, tokens.tobytes())
+
+        key_prefix='sentence:'
+        sentence_key=remove_prefix(record['key'],key_prefix)
+        token_key = f"tokenized:bert:qa:{sentence_key}:{idx}"
+        redisAI.setTensorInKey(token_key, tensor)
+        execute('SADD',f'processed_docs_stage3_tokenized{hash_tag}', token_key)
+
+```
+
+See [full code on github](https://github.com/applied-knowledge-systems/the-pattern-api/blob/156633b9934f1243775671ce6c18ff2bf471c0ce/qasearch/tokeniser_gears_redisai.py#L17)
+
+Then for each Redis Cluster shard, we pre-load BERT QA model by downloading, exporting it into torchscript, then loading it to each shard:
+
+```python
+def load_bert():
+    model_file = 'traced_bert_qa.pt'
+
+    with open(model_file, 'rb') as f:
+        model = f.read()
+    startup_nodes = [{"host": "127.0.0.1", "port": "30001"}, {"host": "127.0.0.1", "port":"30002"}, {"host":"127.0.0.1", "port":"30003"}]
+    cc = ClusterClient(startup_nodes = startup_nodes)
+    hash_tags = cc.execute_command("RG.PYEXECUTE",  "gb = GB('ShardsIDReader').map(lambda x:hashtag()).run()")[0]
+    print(hash_tags)
+    for hash_tag in hash_tags:
+        print("Loading model bert-qa{%s}" %hash_tag.decode('utf-8'))
+        cc.modelset('bert-qa{%s}' %hash_tag.decode('utf-8'), 'TORCH', 'CPU', model)
+        print(cc.infoget('bert-qa{%s}' %hash_tag.decode('utf-8')))
+```
+
+Full [code in github](https://github.com/applied-knowledge-systems/the-pattern-api/blob/156633b9934f1243775671ce6c18ff2bf471c0ce/qasearch/export_load_bert.py)
+
+And when question comes from the user we tokenize and append question to potential answer before:
```

Comment:
```suggestion
And when a question comes from the user we tokenize and append question to potential answer before:
```

Reply:
This doesn't make sense - not sure how to fix it as I don't know what the appending part is intended to mean.